### PR TITLE
April fools maps

### PR DESCRIPTION
**Pull request by RedRafe**
This would be the 1st draft of the April Fools scenario.

As of current time of posting this. I'm running out of time before 2024 April Fools', so I expect this PR to be a draft/test of the features rather than a finished product. Some tips to navigate the new code:

**Relevant global redmew changes**
- `resources/alien_biomes` : handles compatibility if Ab is enabled (needs to be imported in scenarios to work)
- `map_gen/shared/alien_biomes` : generates surfaces MapGenSettings on request (using `resources/alien_biomes`)
> the Alien Biomes modules are shared with CrashSiteV2 draft as well
- `luackeckr` : added couples fields. `script.on_event` was missing (maybe on purpose? seems quite relevant) but I had to add it because our `event` module does NOT handle `CustomInput` events (crashes on init), which are instead supported with standard `core/event_handler` so I'll probably update our Event module when I'll have time / if it'll be needed
- `utils/gui_test` : crashed in some occasions, IDK the exact reason, will need to investigate later

**Main dir**: `map_gen/maps/april_fools` is the new subfolder, structured as:
--- `scenario` : relevant features that may or may not be included in the map
--- `modules` : only "random/crazy" modules (each module is a 1-N level of some crazy behaviour, controlled by the scenario)
--- `2019.lua` and `2024.lua` : the two version of the scenario (reconstructed from `Pinguin` of 2019 and expanded)

There's a good degree of copy/paste to quickly adapt some of our modules to work outside RedmewSurface, I'll tidy 'em up if needed afterwards. Half of the scenario is at [RedMew Data/april-fools](https://github.com/RedRafe/redmew-data/tree/main/april-fools) (for the data stage part and some minor tweaks)

**How to test**
- download & enable *Alien Biomes* mod
- add RedMew-Data mod v0.2.4+ (below)
- scenario is already in debug mode so you'll get some starting items to test around as well
- useful commands: `/af-debug`, `/af-reset`, `/af-max`, & the 3 shortcuts in your quickbar: `teleport up`, `Teleport down`, `Parallel camera`

I'm attaching below a copy of my mod-setting.dat, v0.2.4 of the RedMew mod, & a copy of a save

**Short-term feature until event launches**
- [x] Maybe tweak market's items & costs
- [x] ~~Add a RPG system to increase mininig speed and combat stats for players (RedMew)~~ using DoingThingsByHand mod
- [x] Add default biters on `island` or `mines` or both (RedMew)
- [x] Add iron poles cuz no wood (RedMew Data)
- [x] Duplicate tech tree for more triggers (RedMew Data)
- [x] Add tunnel tech/recipes proper unlocks (RedMew Data)
- [x] Add power link to tunnels between surfaces (RedMew Data)
- [x] ~~Add/Fix/Refactor `scenario/nice-fill.lua` feature (RedMew)~~ using NiceFill mod
- [x] Rework resources generation/frequency/richness to be similar to Diggy (RedMew)
- [ ] Add supports & cave-ins for Mines (RedMew Data/RedMew)
- [x] Add win condition/N. satellites launched 
- [x] Add worms to Mines map gen
- [x] Add preset/poll/restart commands

[mod-settings.dat.zip](https://github.com/Refactorio/RedMew/files/14782610/mod-settings.dat.zip)
[redmew-data_0.2.4.zip](https://github.com/Refactorio/RedMew/files/14782612/redmew-data_0.2.4.zip)
[af_test-deflated.zip](https://github.com/Refactorio/RedMew/files/14782628/af_test-deflated.zip)

(yup, fails build because of `nice-fill.lua` I know)

### CODE REVIEW

File: map_gen/maps/april_fools/2024.lua [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+  - Alien Biomes v0.6.8
+  - RedMew Data  v0.2.4
+]]
+
+local ScenarioInfo = require 'features.gui.info'
+ScenarioInfo.set_map_name('Double Trouble')
+ScenarioInfo.set_map_description('You are Pinguins in Antarctica and Miners underground!')
+ScenarioInfo.set_map_extra_info('Watch out for Icebergs!')
+
+_DEBUG = true
```
Comment (grilledham) [inline map_gen/maps/april_fools/2024.lua]: Probably don't want debug to be true when you launch the map. I typically set `_DEBUG = true` in my `map_selection.lua` file as that file isn't in source control.

**Review by grilledham — APPROVED** [review]
Took a very quick scan through the code and seems mostly fine. I'm happy to merge as is, but there are some changes you could make if you have time, or we can do later.

1. luacheck is failing. I see you mentioned this. It would be nice to fix though. Is there a reason `nice-fill.lua` uses global (as in not local) functions and variables?
2. Do you have the images that were used for the files in `data/presets`. If so could you add them to `map_gen\data\.source_images`.
3. I'd be happy for you to add these scenarios to scenario templates. Would make it easier  for admins to restart a fresh map if needed.

To answer some of your questions:
- The event module not supporting CustomInput events. It fine to add `script.on_event` to .luacheckrc, though it probably was left out deliberately, it's not a big deal. I think the event module predates custom events which is why it doesn't support them, so feel free to add support for them if you want to.
- utils/gui_test crashing. I'm guessing you either added code to add UI elements to the top, or a mod does. Either way the change to the test code is good.

**Comment by RedRafe** [comment]
Hi, thanks a lot for reviewing with such short notice. I'm still trying to add some features of the bullet list before tomorrow (so wait a sec before merging :-) ). Just to answer your points:

1. nicefill code was taken from nicefill mod, I dont have time to refactor it for us so i'll be ditching the module and just add the mod to the event modlist. but I'll be happy to implement one for us in the future
2. I do not have the images unfortunately, only the code
3. will add scenarios to templates & turn off debug mode obviously (and also add poll/restart module to them too I guess)

4. `script.on_event` I really don't like to add to luacheck cuz I also think it was left out to catch `script.` instead of using the event. I'm going to disable the check in that line instead of adding to luacheck, and implement the customInput support later
5. I was getting crashes before adding any gui, don't know it it was a mod or something related to the toast module...

**Comment by RedRafe** [comment]
@grilledham I believe this to be the "final" version before the event. 
Got a couple features in from the bullet list too. I would have loved to add Diggy's cave collapses to the underworld surface, but need to rework the Diggy's module to work outside surface 'redmew' and check other Diggy's template data... I'll have a look tomorrow if I can

Other than that, maybe just have a look at `../scenario/restart_command.lua` (which is the main addition that uses `shared_globals.lua` and `rocket_launched.lua` to set the target condition & pass it along) if you want

**Comment by grilledham** [comment]
LGTM. Would you like me to merge now?

**Comment by RedRafe** [comment]
Sure

**Pull request merged**